### PR TITLE
support filesystems

### PR DESCRIPTION
**Pull request by tornado404**

### CODE REVIEW

File: pkg/apiserver/common/utils.go [hunk 1]
```diff
@@ -123,6 +123,11 @@ func ID(userName, fsName string) string {
 	return FsPrefix + userName + "-" + fsName
 }
 
+// PVCName generate PVC name by fsID
+func PVCName(fsID string) string {
```
Comment (D0m021ng) [inline pkg/apiserver/common/utils.go]: Please refer to  [ConcatenatePVCName](https://github.com/PaddlePaddle/PaddleFlow/blob/develop/pkg/common/schema/fs.go#L59)

File: pkg/apiserver/controller/job/job.go [hunk 2]
```diff
@@ -697,14 +699,37 @@ func getRuntimeByQueue(ctx *logger.RequestContext, queueID string) (runtime.Runt
...
+			return fmt.Errorf("find file system %s failed, err: %v", fs.Name, err)
+		}
 	}
+
 	return nil
 }
+
+func fixFsInConf(conf *schema.Conf, jobSpec *JobSpec, userName string) {
+	if jobSpec.FileSystem.Name != "" {
+		fsID := common.ID(userName, jobSpec.FileSystem.Name)
```
Comment (D0m021ng) [inline pkg/apiserver/controller/job/job.go]: It might be better to call fs api.

File: pkg/apiserver/controller/job/job.go [hunk 3]
```diff
@@ -697,14 +699,37 @@ func getRuntimeByQueue(ctx *logger.RequestContext, queueID string) (runtime.Runt
 	return runtimeSvc, nil
 }
 
-func validateFileSystem(fs schema.FileSystem, userName string) error {
-	if fs.Name == "" {
+func validateFileSystem(jobSpec *JobSpec, userName string) error {
```
Comment (D0m021ng) [inline pkg/apiserver/controller/job/job.go:701]: Merge it with `fixFsInConf`

File: pkg/job/runtime/kubernetes/executor/kubernetes_job.go [hunk 4]
```diff
@@ -385,9 +382,10 @@ func (j *KubeJob) fillContainerInTasks(container *corev1.Container, task models.
 		container.Args = task.Args
 	}
 	container.Resources = j.generateResourceRequirements(task.Flavour)
-	if j.VolumeName != "" {
-		container.VolumeMounts = j.appendMountIfAbsent(container.VolumeMounts, j.generateVolumeMount())
-	}
+
+	taskFs := task.Conf.GetAllFileSystem()
```
Comment (D0m021ng) [inline pkg/job/runtime/kubernetes/executor/kubernetes_job.go:385]: What happens when `taskFs` is empty.

File: pkg/job/runtime/kubernetes/executor/kubernetes_job.go [hunk 5]
```diff
@@ -733,3 +733,109 @@ func (j *KubeJob) getDefaultTemplate(framework schema.Framework) ([]byte, error)
 	}
 	return extConf, nil
 }
+
+func generateVolumes(fileSystem []schema.FileSystem) []corev1.Volume {
+	var vs []corev1.Volume
+	if len(fileSystem) == 0 {
+		log.Debug("len(fileSystem) == 0")
```
Comment (D0m021ng) [inline pkg/job/runtime/kubernetes/executor/kubernetes_job.go]: Add more log info.

File: pkg/job/runtime/kubernetes/executor/kubernetes_job.go [hunk 6]
```diff
@@ -733,3 +733,109 @@ func (j *KubeJob) getDefaultTemplate(framework schema.Framework) ([]byte, error)
...
+		return vs
+	}
+
+	for _, fs := range fileSystem {
+		log.Debugf("FileSystems[%+v]", fs)
+		volume := corev1.Volume{
```
Comment (D0m021ng) [inline pkg/job/runtime/kubernetes/executor/kubernetes_job.go:747]: Remove duplicated volumes?

File: pkg/job/runtime/kubernetes/executor/kubernetes_job.go [hunk 7]
```diff
@@ -733,3 +733,109 @@ func (j *KubeJob) getDefaultTemplate(framework schema.Framework) ([]byte, error)
...
+		}
+		vs = append(vs, volume)
+	}
+
+	return vs
+}
+
+func generateVolumeMounts(fileSystems []schema.FileSystem) []corev1.VolumeMount {
+	var vms []corev1.VolumeMount
+	if len(fileSystems) == 0 {
```
Comment (D0m021ng) [inline pkg/job/runtime/kubernetes/executor/kubernetes_job.go:764]: log

File: pkg/apiserver/controller/job/job.go (hunk 2 above)
Comment (tornado404) [reply pkg/apiserver/controller/job/job.go]: 这个方法已经被删掉了

File: pkg/apiserver/controller/job/job.go (hunk 3 above)
Comment (tornado404) [reply pkg/apiserver/controller/job/job.go:701]: ok

File: pkg/job/runtime/kubernetes/executor/kubernetes_job.go (hunk 4 above)
Comment (tornado404) [reply pkg/job/runtime/kubernetes/executor/kubernetes_job.go:385]: return nil slice and happens nothing

File: pkg/job/runtime/kubernetes/executor/kubernetes_job.go (hunk 5 above)
Comment (tornado404) [reply pkg/job/runtime/kubernetes/executor/kubernetes_job.go]: ok

File: pkg/job/runtime/kubernetes/executor/kubernetes_job.go (hunk 6 above)
Comment (tornado404) [reply pkg/job/runtime/kubernetes/executor/kubernetes_job.go:747]: Duplicated volumes will be removed when appendVolumesIfAbsent

File: pkg/job/runtime/kubernetes/executor/kubernetes_job.go (hunk 7 above)
Comment (tornado404) [reply pkg/job/runtime/kubernetes/executor/kubernetes_job.go:764]: ok

**Review by D0m021ng — APPROVED** [review]

**Pull request merged**